maven/poms.bzl

```diff
@@ -45,38 +65,223 @@ def _process_dependency(dep_node):
+
+# Extracts dependency coordinates from a given <dependencies> node within a <dependencyManagement> node of a pom node.
+# The parameter should be an xml node containing the tag <dependencyManagement>
+def _extract_dependency_management(project_node):
+    node = xml.find_first(project_node, labels.DEPENDENCY_MANAGEMENT)
+    return poms.extract_dependencies(node) if bool(node) else []
+
+def _process_parent(dep_node):
+    group_id = None
+    artifact_id = None
+    version = None
+    for c in dep_node.children:
+        if c.label == labels.GROUP_ID:
+            group_id = c.content
+        elif c.label == labels.ARTIFACT_ID:
+            artifact_id = c.content
+        elif c.label == labels.VERSION:
+            version = c.content
+    return struct(
+        group_id = group_id,
+        artifact_id = artifact_id,
+        version = version,
+        packaging = "pom",  # Parent POMs must be pure metadata artifacts (only a .pom, no .jar/.aar, etc.)
+        original_spec = "%s:%s:%s:pom" % (group_id, artifact_id, version),
+        classifier = None,
     )
 
-# Extracts dependency coordinates from a given <dependencies> node of a pom node.  The parameter should be the project
-# node of a parsed xml document tree, returned by poms.parse(xml_text)
-def _extract_dependencies(project):
-    dependencies = []
+# Extracts parent specification for the supplied pom file.
+# The parameter should be the project node of a parsed xml document tree, returned by poms.parse(xml_text)
+def _extract_parent(project):
+    for node in project.children:
+        if node.label == labels.PARENT:
+            return _process_parent(node)
+    return None
+
+def _extract_properties(project):
+    properties_nodes = []
     for node in project.children:
-        if node.label == "dependencies":
-            dependencies = node.children
-    return [_process_dependency(x) for x in dependencies]
+        if node.label == labels.PROPERTIES:
+            properties_nodes = node.children
+    properties = {}
+    for node in properties_nodes:
+        properties[node.label] = node.content
+    return properties
 
 def _format_dependency(dep):
     result = "%s:%s:%s" % (dep.group_id, dep.artifact_id, dep.version)
     if bool(dep.classifier):
         type = dep.type if bool(dep.type) else "jar"
         result = "%s:%s" % (result, dep.type)
-    else:
-        if bool(dep.type) and not dep.type == "jar":
-            result = "%s:%s" % (result, dep.type)
+    elif bool(dep.type) and not dep.type == "jar":
+        result = "%s:%s" % (result, dep.type)
     return result
 
 def _parse(xml_text):
     root = xml.parse(xml_text)
     for node in root.children:
-        if node.label == "project":
+        if node.label == labels.PROJECT:
             return node
     fail("No <project> tag found in supplied xml: %s" % xml)
 
+# A pom-specific node-merge algorith,
+def _merge_content_last_wins(a, b):
+    if not bool(a):
+        return b
+    elif not bool(b):
```

Comment:
nit: else after return is redundant. Also, could you factor out the return, `return b if not bool(a) elif...`? Not sure if that'd be more readable or less.

Reply:
I had it structured more with the if-as-expression style, and it just read in a very confusing (to this java/kotlin guy) way.  it might be less pythonic, but this way at least was more symmetrical.  The else is redundant, but mostly there to keep indent symmetry.   I'll clean all these up.